tests/unit/core.js

```diff
@@ -183,24 +183,109 @@ exports.testNewNonNativeObject = function (test) {
   test.done();
 };
 
+exports["undefined as variable name"] = function (test) {
 
-/**
- * Test that JSHint allows `undefined` to be a function parameter.
- * It is a common pattern to protect against the case when somebody
- * overwrites undefined. It also helps with minification.
- *
- * More info: https://gist.github.com/315916
- */
-exports.testUndefinedAsParam = function (test) {
-  var code  = '(function (undefined) {}());';
-  var code1 = 'var undefined = 1;';
+  TestRun(test)
+    .test("var undefined;");
 
-  TestRun(test).test(code);
+  TestRun(test)
+    .test("let undefined;", { esnext: true });
+
+  TestRun(test)
+    .addError(1, "const 'undefined' is initialized to 'undefined'.")
```

Comment:
http://www.ecma-international.org/ecma-262/6.0/index.html#sec-let-and-const-declarations-static-semantics-early-errors
```
It is a Syntax Error if Initializer is not present and IsConstantDeclaration of the LexicalDeclaration containing this production is true.
```